予想：女性の方が生き残る確率が高かったはず、等級の高いほうが生き残っている確率が高いはず

- 提供されるデータの変数を見てみたところ、生き残ったかどうかのデータと、部屋の等級と家族関係のデータは提供されてる
- まず、家族連れだったかそうでないかのデータを整形→等級をわけるが1つ目の整理か。そして、おそらく妥当なのは女性と子供は優先的に救命ボートに乗せられただろうと想定し、それは等級が高い順であったであろうと予測。データを出してみてそれが本当かどうか見てみて判断って感じかなあ？（でもな、家族連れのデータを整形して見比べるってどうやるんだろう。最初にやるにはなんか手間が多い気がする。もっと何かある気がする）
- （当時の社会的背景とか文化を考えると）おそらく生き残りやすいであろう順番は、女の子供 ＞ 男の子供 ＞ 家族連れの女性 ＞ 単独の女性 ＞家族連れの男性＞単独の男性 ＞乗務員
- より詳細にすると女の子供等級が高い順 ＞ 男の子供等級が高い順 ＞ 家族連れの女性等級が高い順 ＞ 単独の女性等級が高い順 ＞家族連れの男性等級が高い順＞単独の男性等級が高い順 ＞乗務員
- あとは子供の定義だよね。タイタニック号の沈没は1912年まあ20世紀だとはいえ、成人とかっていう定義がわからない
- 子供については家族関係のデータを紐づける必要はないから、女の子供をまず等級別にならべて、男の子供を等級別にならべるのが一番最初だね。そのあと、女性が既婚か未婚かで並べて、つぎは男性の社会的地位と家族持ちかをしらべて、最後に独身男性かな

In [1]:
import numpy as np # numpy を np という略称でインポート
import pandas as pd # Pandas を pd という略称でインポート

#import os
#for dirname, _, filenames in os.walk('/kaggle/input'):
#    for filename in filenames:
#        print(os.path.join(dirname, filename))

# =========================
# データの読み込みと確認
# =========================

train_data = pd.read_csv("/kaggle/input/titanic/train.csv")

# データの一部を表示して、正しく読み込めたか確認
#print(train_data.head(5))

# 変数名(列名)を調べる
print(train_data.columns)

# ==================================
# 仮説：女性の方が生存率が高いのでは？
# ==================================

# 性別ごとの生存率
print("----- 性別ごとの生存率 -----")
# Sex を groupby するので male と female に分かれる
# グループ化したデータのうち Survived というラベルの列だけを選択する
# mean で各グループの Survived の平均を計算する
print(train_data.groupby('Sex')['Survived'].mean())

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')
----- 性別ごとの生存率 -----
Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


予想通り、女性の生存率の平均が0.74、男性の生存率が、0.19。

**NumPy**
- 数値計算するためのライブラリ

**Pandas**
- データ分析のためのライブラリ

**pandas.read_csv()**
> Read a comma-separated values (csv) file into DataFrame.
> https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

DateFrameとして格納される

**pandas.DataFrame.columns**
> The column labels of the DataFrame.
> https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html#pandas.DataFrame.columns

DateFrameの列のラベルが見られる

**pandas.DataFrame.groupby**

> Group DataFrame using a mapper or by a Series of columns.
> https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html#pandas.DataFrame.groupby

データフレームを一次元の列のによってグループ化する

📝 SQLっぽい

❓️ mapperってなんだろう

- TRPG だと地図を書く人
- 各行=レコードを見て→指定した変数（列）をチェック→なんの値が入っているかを確認→その値に基づいてレコードをグループに分ける役割

**pandas.core.groupby.DataFrameGroupBy.mean**

> Compute mean of groups, excluding missing values.
> https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.mean.html#pandas.core.groupby.DataFrameGroupBy.mean

❓️ 欠損値ってなんだろう

> 欠損値とは、データ・セット中の不明な値、収集されていない値、または誤って入力された値です。通常、そのような値はフィールドに対して無効になります。
>https://www.ibm.com/docs/ja/spss-modeler/saas?topic=values-overview-missing


📝 **Survived の平均 = 生存者の割合**

- Survived は 0 と 1 しか値をとらないので全部足して全人数で割ると生存者の割合が出る。
- 標本比率 xバー は標本平均 p ハットと同じのやつ

In [2]:
# ==========================================
# 仮説：社会的地位が高い方が生存率が高いのでは？
# ==========================================

# 等級（Pclass）ごとの生存率
print("-----等級ごとの生存率 -----")
print(train_data.groupby('Pclass')['Survived'].mean())

-----等級ごとの生存率 -----
Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


等級も

1等 0.629630
2等 0.472826
3等 0.242363

で、等級が高い（社会的地位が高い）で予想通り

**ここまでの分析**

- 男性より女性が生き残っているのは、女性を優先するという文化によって救命ボートに乗る確率が高かったから
- 客室の等級が高いほうが生き残っているのは、社会的地位というものが重視される時代であり、救命ボートに乗れる確率が高かったから

次の予想：子供のほうが大人より生き残り確率が高かったはず

In [3]:
# ==============================================
# 仮説：子供のほうが大人より生き残り確率が高いのでは？
# ==============================================

# 年齢ごとの生存率
print("年齢ごとの生存率:")
print(train_data.groupby('Age')['Survived'].mean())

年齢ごとの生存率:
Age
0.42     1.0
0.67     1.0
0.75     1.0
0.83     1.0
0.92     1.0
        ... 
70.00    0.0
70.50    0.0
71.00    0.0
74.00    0.0
80.00    1.0
Name: Survived, Length: 88, dtype: float64


ざっとすべての年齢の生き残り平均をだしてみたら、一部しか表示されなかったけど、面白かったのは、やはり数字が小さければ小さいほど生存確率は1で、最後の方は80.00だけ1なのは、きっと、敬老なんだろう

In [4]:
# ==============================================
# 性別と等級を組み合わせて、生存率をチェックする
# ==============================================

# 性別（'Sex'）と等級（'Pclass'）でグループ化し、'Survived'の平均を計算
compound_survival = train_data.groupby(['Pclass', 'Sex'])['Survived'].mean()

# 結果を表示
print("--- 等級と性別を組み合わせた生存率 ---")
print(compound_survival)

--- 等級と性別を組み合わせた生存率 ---
Pclass  Sex   
1       female    0.968085
        male      0.368852
2       female    0.921053
        male      0.157407
3       female    0.500000
        male      0.135447
Name: Survived, dtype: float64


面白い結果が出た。等級1と2の女性の生存率は0.96と0.92でほぼ変わらないが等級3になったとたん、女性でも0.5になる。船室の配置がどういう状況だったか調べれば、答えが出そう

In [5]:
# ======================================================
# 15歳以下を子供と仮定して、等級と性別ごとの生存率をチェックする
# ======================================================

# 年齢15歳以下を子供と仮定
print("--- 年齢15歳以下を子供と仮定 ---")
children = train_data[train_data['Age'] <= 15] 

# Pclass（等級）と Sex（性別）でグループ化し、生存率（mean）と人数（count）を同時に計算
children_stats = children.groupby(['Pclass', 'Sex'])['Survived'].agg(['mean', 'count'])

# 結果を表示
print("--- 子供の等級・性別ごとの生存率と人数 ---")
print(children_stats)

--- 年齢15歳以下を子供と仮定 ---
--- 子供の等級・性別ごとの生存率と人数 ---
                   mean  count
Pclass Sex                    
1      female  0.666667      3
       male    1.000000      3
2      female  1.000000     10
       male    1.000000      9
3      female  0.533333     30
       male    0.321429     28


/usr/local/lib/python3.11/dist-packages/pandas/core/computation/expressions.py:73: RuntimeWarning: invalid value encountered in less_equal
  return op(a, b)


なんでだろう、年齢15歳以下を子供と仮定すると、等級1のfemaleの子供は0.6、等級2は男女ともに1だ

おそらくなにか外れ値があるのだろうと予想したが、人数が少なかった可能性がある。ということは、まずこのデータを出す前に、それぞれの子供の数を出力しておかないとね

コードにcountを追加した

1等級の子供は男女ともに3人だけだった。おそらく1人がなんらかの理由で亡くなったことにより、1等級の女の子供の生存平均が減ったんだろうね

結論：
1等客室の女の子の生き残りが少ないのは、そもそも子供の数が少なかったので1人が亡くなっただけでも死亡率が上がる状態

**pandas.core.groupby.DataFrameGroupBy.agg**
> Aggregate using one or more operations over the specified axis.
> https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.agg.html#pandas.core.groupby.DataFrameGroupBy.agg

複数の集計操作を同時に実行する。

集計関数を

.agg(['mean', 'count']

みたいに書くと平均を出しつつ、1等級でかつ女性の総数、みたいなのを出してくれる

なぜ、3等客室の女性の予測に比べて生存率が低いのか

In [6]:
# =====================================
# 疑問：3等の女性の生存率はなぜ低いのだろう？
# =====================================

# 'Pclass'が3 かつ 'Sex'が'female'の行を抽出
third_class_women = train_data[(train_data['Pclass'] == 3) & (train_data['Sex'] == 'female')]

# 生存率と人数を計算
third_class_women_stats = third_class_women['Survived'].agg(['mean', 'count'])

print("--- 3等船室の女性の生存率と人数 ---")
print(third_class_women_stats)

--- 3等船室の女性の生存率と人数 ---
mean       0.5
count    144.0
Name: Survived, dtype: float64


3等船室の女性の人数が144人と多いにも関わらず、生存率が 0.50 に半減している

countは144だから、数が少ないからという理由ではないことがわかる。ということは、逃げられなかったんだろう

ウィキペディアで調べたところ、3等客室の女性の部屋は船尾側にあったらしい。家族連れも船尾に部屋があるみたい。タイタニック号は勝手に真ん中からぱっくり割れたイメージがあったから、真ん中に持ち上がることを考えると船尾と船首側は先に沈んだのだろうか

あとは「子供」がどこからかがわからないのを解決したい。Name のデータにMissとかMrとかを持っているからそれを見てみる

In [7]:
# =========================
# 敬称から生存確率を調べてみる
# =========================

# Name列から敬称を抽出
# Caldwell, Mr. Albert Francis みたいなデータだから
# スペースの後に英字が続き、ピリオドで終わる部分」を抜き出す
train_data['Title'] = train_data['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)

# 敬称（Title）でグループ化し、生存率と人数を同時に計算
title_stats = train_data.groupby('Title')['Survived'].agg(['mean', 'count'])

# 結果を表示
print("--- 敬称ごとの生存率と人数 ---")
print(title_stats)

--- 敬称ごとの生存率と人数 ---
              mean  count
Title                    
Capt      0.000000      1
Col       0.500000      2
Countess  1.000000      1
Don       0.000000      1
Dr        0.428571      7
Jonkheer  0.000000      1
Lady      1.000000      1
Major     0.500000      2
Master    0.575000     40
Miss      0.697802    182
Mlle      1.000000      2
Mme       1.000000      1
Mr        0.156673    517
Mrs       0.792000    125
Ms        1.000000      1
Rev       0.000000      6
Sir       1.000000      1


train_data['Title']で すでにあるDataFrame に新しい列を作ってそこに抽出したかったデータを入れている

配列に入れなくても抽出してまとめてくれるの便利すぎる

- SQLみたいに抽出したテーブルを作ってる感じかと思われる
- これができるのは、train_dataっていう変数のなかに、pandasを利用してcsvをdateFrameとして入れたからなはず

私の予想は、女性と子供の生存率は高いだろうとものだった
女性や子供の敬称でも、人数の少ないものはいったん除外して考えるとして

**女性**
- Miss（未婚女性）約半数。人数が多いので正確性は高そう
- Mlle（未婚女性）生存率100％。ただし、人数が少ない
- Mme（婦人）生存率100％。ただし、人数が少ない
- Mrs（既婚女性）生存率約8割。人数も十分か
- Ms （女性）生存率100％。ただし、人数が少ない
- Countess（伯爵夫人）生存率100％。ただし、人数が少ない
- Lady（貴族の妻？）生存率100％。ただし、人数が少ない

**男性**
- Master（若い男性）生存率約6割。40人なのでまあまあの人数
- Mr（男性）生存率約１割。人数も十分
- その他の男性は生存していても人数が少ないのでいったんおいておく


これで予想は立った

まずは、ごくごく簡単な、女性なら1（生存）男性なら0（死亡）というプログラムを作成する

In [8]:
# ==================================
# 一番小さいモデルを作ってみる
# ==================================

# 1. テストデータを読み込む
test_data = pd.read_csv("/kaggle/input/titanic/test.csv")

# 2. 全員をいったん「死亡（0）」とする
test_data['Survived'] = 0

# 3. 性別が「女性（female）」の乗客を「生存（1）」にする
test_data.loc[test_data['Sex'] == 'female', 'Survived'] = 1

# 4. 提出用のデータフレームを作成（PassengerIdとSurvivedのみ）
submission = test_data[['PassengerId', 'Survived']]

# 5. CSVファイルとして保存（Kaggleに提出するファイル）
submission.to_csv('gender_submission.csv', index=False)

print("提出ファイル (gender_submission.csv) が作成されました")

提出ファイル (gender_submission.csv) が作成されました


**pandas.DataFrame.loc**
> Access a group of rows and columns by label(s) or a boolean array.
> https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html

ラベルか真偽配列に基づいて、行と列のグループにアクセスする

具体的に言うと下記のような感じ

- Sex が female になっている行をすべて選ぶ
- 選んだ行の、Survived の列をロックオンする
- 上記条件が整ったら、選択されたすべての Survived のセルに1を入れる
